src/main/scala/org/apache/spark/ml/feature/RapidsPCA.scala

```diff
@@ -168,26 +171,71 @@ class RapidsPCAModel(
    */
   override def transform(dataset: Dataset[_]): DataFrame = {
     val outputSchema = transformSchema(dataset.schema, logging = true)
+    val gpuIdBC = dataset.rdd.context.broadcast(getGpuId)
 
     // TODO(rongou): make this faster and re-enable.
-    //    if (getUseGemm) {
-    //      val transformed = dataset.toDF().rdd.mapPartitions(iterator => {
-    //        val gpuID = TaskContext.get().resources()("gpu").addresses(0)
-    //        val partition = iterator.toList
-    //        val A = Matrices.fromVectors(partition.map(_.getAs[Vector]($(inputCol)))).toDense
-    //        val C = Matrices.zeros(partition.length, getK).toDense
-    //        CUBLAS.gemm_b(A, pc, C, gpuID)
-    //        C.rowIter.zip(partition.iterator).map { case (v, r) =>
-    //          Row.fromSeq(r.toSeq ++ Seq(v))
-    //        }
-    //      })
-    //      dataset.sparkSession.createDataFrame(transformed, outputSchema)
-    //    }
-    //    else {
-    val transposed = pc.transpose
-    val transformer = udf { vector: Vector => transposed.multiply(vector) }
-    dataset.withColumn($(outputCol), transformer(col($(inputCol))), outputSchema($(outputCol)).metadata)
-    //    }
+    if (getUseGemm) {
+      val input = dataset.select($(inputCol)).rdd.map {
+        case Row(v: Vector) => v
+      }
+      val n = input.first().size
+
+      val transformed = input.mapPartitions(iterator => {
+        val gpu = if (gpuIdBC.value == -1) {
+          TaskContext.get().resources()("gpu").addresses(0).toInt
+        } else {
+          gpuIdBC.value
+        }
+        val partition = iterator.toList
+        val bas = partition.map(v => v.asBreeze.toArray)
+        val nvtxRangeConcat = new NvtxRange("concat before transform", NvtxColor.PURPLE)
+
+        val A = try {
+          new DenseMatrix(bas.length, n, Array.concat(bas: _*), isTransposed = true)
```

Comment:
Consider using unsafe memory copy to replace Array.concat. This step takes about 10 seconds in profiling while GPU computation takes 2 seconds